src/bytes.rs

```diff
@@ -1488,7 +1485,7 @@ impl Borrow<[u8]> for BytesMut {
 
 impl fmt::Write for BytesMut {
     fn write_str(&mut self, s: &str) -> fmt::Result {
-        BufMut::put(self, s);
+        self.extend_from_slice(s.as_bytes());
```

Comment:
If a `BytesMut` should be treated like a `&mut [u8]`, then I propose that the `fmt::Write` work the same way.

While there isn't a `fmt::Write` impl for `&mut [u8]`, there is an `io::Write`, (so you can do `write!(slice_mut, "{}", foo)`. The `Write` impl writes all it can to the end of the buffer, and then additional writes return `Ok(0)`. The `io::Write::write_fmt` default method tries to use `write_all`, so it will eventually return a `WriteZero` error if the mutable slice doesn't have any more room. In turn, this is translated to `fmt::Error`. This means that it can error, but it **doesn't panic** unless you choose to, by unwrapping the error.

Proposed implementation:

```rust
impl fmt::Write for BytesMut {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        if s.len() <= self.remaining_mut() {
            // hoping that the compiler can elide the assert in `put_slice`
            // since it's the same check
            self.put_slice(s.as_bytes());
            Ok(())
        } else {
            Err(fmt::Error)
        }
    }
}
```